### ISSUE TITLE
[JENKINS-54010] Support the visualization of two levels of parallelity in stages

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#4878]
<p>Support scripted parallel blocks in parallel stages in a declarative pipeline.</p>

<p>Example pipeline:</p>

<pre>
pipeline {
    agent none
    stages {
stage(<span class="code-quote">'Parallel Stage'</span>) {
    parallel {
stage(<span class="code-quote">'Stage 1'</span>) {
    steps {
echo <span class="code-quote">"Stage 1"</span>
    }
}
stage(<span class="code-quote">'Stage 2'</span>) {
    steps {
script {
    parallel (
<span class="code-quote">"Stage 2.1."</span>: {
    echo <span class="code-quote">"Stage 2.1."</span>
},
<span class="code-quote">"Stage 2.2."</span>: {
    echo <span class="code-quote">"Stage 2.2."</span>
}
    )
}
    }
}
    }
}
    }
}
</pre>
<p>Currently it is not properly visualized in blue ocean as shown in parallel-stages.png</p>

<p> </p>

<p>The idea how it should be visualized is shown in parallel-stages-new.png</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kurzy">kurzy</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010" target="_blank">Support the visualization of two levels of parallelity in stages</a></i></summary>
<i><ul>
<li><b>status</b>: In Progress
<li><b>priority</b>: Blocker
<li><b>component(s)</b>: blueocean-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 133
<li><b>watchers</b>: 141
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Support scripted parallel blocks in parallel stages in a declarative pipeline.</p>

<p>Example pipeline:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
    agent none
    stages {
        stage(<span class="code-quote">'Parallel Stage'</span>) {
            parallel {
                stage(<span class="code-quote">'Stage 1'</span>) {
                    steps {
                        echo <span class="code-quote">"Stage 1"</span>
                    }
                }
                stage(<span class="code-quote">'Stage 2'</span>) {
                    steps {
                        script {
                            parallel (
                                <span class="code-quote">"Stage 2.1."</span>: {
                                    echo <span class="code-quote">"Stage 2.1."</span>
                                },
                                <span class="code-quote">"Stage 2.2."</span>: {
                                    echo <span class="code-quote">"Stage 2.2."</span>
                                }
                            )
                        }
                    }
                }
            }
        }
    }
}
</pre>
</div></div>
<p>Currently it is not properly visualized in blue ocean as shown in parallel-stages.png</p>

<p> </p>

<p>The idea how it should be visualized is shown in parallel-stages-new.png</p></pre>
</details>
</details>
<details><summary><i>9 attachments</i></summary>

- [Screen Shot 2019-05-14 at 9.49.06 AM.png](https://issues.jenkins.io/secure/attachment/47427/Screen%20Shot%202019-05-14%20at%209.49.06%20AM.png)
  > ![Screen Shot 2019-05-14 at 9.49.06 AM.png](https://issues.jenkins.io/secure/attachment/47427/Screen%20Shot%202019-05-14%20at%209.49.06%20AM.png)
- [Screen Shot 2019-05-20 at 8.55.46 PM.png](https://issues.jenkins.io/secure/attachment/47428/Screen%20Shot%202019-05-20%20at%208.55.46%20PM.png)
  > ![Screen Shot 2019-05-20 at 8.55.46 PM.png](https://issues.jenkins.io/secure/attachment/47428/Screen%20Shot%202019-05-20%20at%208.55.46%20PM.png)
- [Selection_090.png](https://issues.jenkins.io/secure/attachment/49809/Selection_090.png)
  > ![Selection_090.png](https://issues.jenkins.io/secure/attachment/49809/Selection_090.png)
- [Selection_093.png](https://issues.jenkins.io/secure/attachment/49810/Selection_093.png)
  > ![Selection_093.png](https://issues.jenkins.io/secure/attachment/49810/Selection_093.png)
- [TwoPipelinesInParallel.png](https://issues.jenkins.io/secure/attachment/53591/TwoPipelinesInParallel.png)
  > ![TwoPipelinesInParallel.png](https://issues.jenkins.io/secure/attachment/53591/TwoPipelinesInParallel.png)
- [TwoPipelinesInSerial.png](https://issues.jenkins.io/secure/attachment/53590/TwoPipelinesInSerial.png)
  > ![TwoPipelinesInSerial.png](https://issues.jenkins.io/secure/attachment/53590/TwoPipelinesInSerial.png)
- [image-2019-12-18-10-50-54-224.png](https://issues.jenkins.io/secure/attachment/49808/image-2019-12-18-10-50-54-224.png)
  > ![image-2019-12-18-10-50-54-224.png](https://issues.jenkins.io/secure/attachment/49808/image-2019-12-18-10-50-54-224.png)
- [parallel-stages-new.png](https://issues.jenkins.io/secure/attachment/44753/parallel-stages-new.png)
  > ![parallel-stages-new.png](https://issues.jenkins.io/secure/attachment/44753/parallel-stages-new.png)
- [parallel-stages.PNG](https://issues.jenkins.io/secure/attachment/44751/parallel-stages.PNG)
  > ![parallel-stages.PNG](https://issues.jenkins.io/secure/attachment/44751/parallel-stages.PNG)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>philsegal:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=351102&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-351102">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I would hope this would also support the same for scripted pipelines too.</p>
</pre>
</details>

</ul>
</details>

<p>I would hope this would also support the same for scripted pipelines too.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cloudcompany:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=351610&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-351610">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>+100 <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> Today, I am stucked to run different kind of tests (python, qtest, ...) in parallel on two configurations (Windows and Linux) in parallel.</p>
</pre>
</details>

</ul>
</details>

<p>+100 <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> Today, I am stucked to run different kind of tests (python, qtest, ...) in parallel on two configurations (Windows and Linux) in parallel.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>digaus:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=353945&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-353945">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This picture demonstrates it better... running parallel within parallel without a step before makes no sense since you could just run them in the outer parallel task.</p>

<p><a href="https://issues.jenkins-ci.org/secure/attachment/36602/pipeline_mockup.png" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/secure/attachment/36602/pipeline_mockup.png</a></p>
</pre>
</details>

</ul>
</details>

<p>This picture demonstrates it better... running parallel within parallel without a step before makes no sense since you could just run them in the outer parallel task.</p>

<p><a href="https://issues.jenkins-ci.org/secure/attachment/36602/pipeline_mockup.png" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/secure/attachment/36602/pipeline_mockup.png</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>asg:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=357541&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-357541">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=digaus" class="user-hover" rel="digaus">digaus</a> Actually the complete idea from Ben Dean would be really nice to have. Check this comment here: </p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-38442?focusedCommentId=292647&amp;page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292647" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-38442?focusedCommentId=292647&amp;page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292647</a></p>

<p>It would be nice to have that not only for the declarative, but also the scripted pipeline.</p>

<p>Sometimes the test suites to cover are not statically declared and can depend on different things for different branches, hens the need for the scripted pipeline.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=digaus" class="user-hover" rel="digaus">digaus</a> Actually the complete idea from Ben Dean would be really nice to have. Check this comment here: </p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-38442?focusedCommentId=292647&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292647" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-38442?focusedCommentId=292647&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292647</a></p>

<p>It would be nice to have that not only for the declarative, but also the scripted pipeline.</p>

<p>Sometimes the test suites to cover are not statically declared and can depend on different things for different branches, hens the need for the scripted pipeline.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mirceaalbu.png" title="mirceaalbu's avatar" /> mirceaalbu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=360327&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-360327">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We need this for the scripted pipeline as well.</p>
</pre>
</details>

</ul>
</details>

<p>We need this for the scripted pipeline as well.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rthakkar:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=365365&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-365365">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is there any progress on this task? My team has a highly parallel flow that requires multiple layers of nested parallels. It would make life easier for debugging purposes if this feature were implemented.</p>
</pre>
</details>

</ul>
</details>

<p>Is there any progress on this task? My team has a highly parallel flow that requires multiple layers of nested parallels. It would make life easier for debugging purposes if this feature were implemented.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>freme:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=365418&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-365418">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>same here - i have a 50+ different build configs to run in parallel, a dozen per OS (Linux, MacOS, iOS, Android, Windows . . .) - Blue Ocean graph is pretty useless this way . . .</p>
</pre>
</details>

</ul>
</details>

<p>same here - i have a 50+ different build configs to run in parallel, a dozen per OS (Linux, MacOS, iOS, Android, Windows . . .) - Blue Ocean graph is pretty useless this way . . .</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>shahafd:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=366197&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-366197">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This feature will help us a lot.</p>

<p>Multiple levels of parallelizm are currently supported only by creating new jobs, which is a poor solution given that Jenkins groovy supports it out of the box.</p>

<p>I'm sure this feature will help a lot of users since many applications test themselves on multiple platforms today.</p>
</pre>
</details>

</ul>
</details>

<p>This feature will help us a lot.</p>

<p>Multiple levels of parallelizm are currently supported only by creating new jobs, which is a poor solution given that Jenkins groovy supports it out of the box.</p>

<p>I'm sure this feature will help a lot of users since many applications test themselves on multiple platforms today.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>asavanchuk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=366208&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-366208">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shahafd" class="user-hover" rel="shahafd">shahafd</a> I'm not sure that you are talking about the BlueOcean plugin.</p>

<p>Jenkins pipeline already supports multiple levels of parallelism, this is only about Blue Ocean's UI</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shahafd" class="user-hover" rel="shahafd">shahafd</a> I'm not sure that you are talking about the BlueOcean plugin.</p>

<p>Jenkins pipeline already supports multiple levels of parallelism, this is only about Blue Ocean's UI</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>shahafd:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=366218&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-366218">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=asavanchuk" class="user-hover" rel="asavanchuk">asavanchuk</a> - I am talking about the BlueOcean plugin.</p>

<p>We work with BlueOcean and we have multiple levels of parallelizm in our pipeline, and the only method to show it correctly using the GUI is creating multiple jobs for the different paltforms, otherwise it shows all of the parallel steps in the internal parallel block in the same level of the parallel steps of the external parallel block which is obviously wrong, furthermore, it also shows the serial steps of the internal job as parallel.</p>

<p> </p>

<p>For example - we have the following pipeline:</p>

<p>Parallel:</p>

<p>    Linux:</p>

<p>        Build</p>

<p>        Parallel:</p>

<p>            Fixed tests</p>

<p>            Custom tests</p>

<p>    Windows:</p>

<p>        Build</p>

<p> </p>

<p>We currently create a different job for windows and linux pipes and trigger them from the master job.</p>

<p>When trying to run it in a single job, BlueOcean shows it as follows:</p>

<p>Parallel:</p>

<p>    Linux</p>

<p>    Fixed tests</p>

<p>    Custom tests</p>

<p>    Windows</p>

<p> </p>

<p>I would very much like to see it as shown in <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> (pipeline_mockup.png)</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=asavanchuk" class="user-hover" rel="asavanchuk">asavanchuk</a> - I am talking about the BlueOcean plugin.</p>

<p>We work with BlueOcean and we have multiple levels of parallelizm in our pipeline, and the only method to show it correctly using the GUI is creating multiple jobs for the different paltforms, otherwise it shows all of the parallel steps in the internal parallel block in the same level of the parallel steps of the external parallel block which is obviously wrong, furthermore, it also shows the serial steps of the internal job as parallel.</p>

<p> </p>

<p>For example - we have the following pipeline:</p>

<p>Parallel:</p>

<p>    Linux:</p>

<p>        Build</p>

<p>        Parallel:</p>

<p>            Fixed tests</p>

<p>            Custom tests</p>

<p>    Windows:</p>

<p>        Build</p>

<p> </p>

<p>We currently create a different job for windows and linux pipes and trigger them from the master job.</p>

<p>When trying to run it in a single job, BlueOcean shows it as follows:</p>

<p>Parallel:</p>

<p>    Linux</p>

<p>    Fixed tests</p>

<p>    Custom tests</p>

<p>    Windows</p>

<p> </p>

<p>I would very much like to see it as shown in <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> (pipeline_mockup.png)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nedrebo:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=367276&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367276">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Our team is also hoping for a solution to this, the setup we have is similar to what <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shahafd" class="user-hover" rel="shahafd">shahafd</a> describes, starting with some sequential pre-work, then splitting into windows and linux build, then ideally further splitting several times down the pipeline (multiple tests that each is parameterized over OS and HW configurations and fully parallelizable).</p>

<p>Right now our pipeline takes three times as long as it should because we really like to have the user friendly visualization we get from Blue Ocean. This is of course not ideal because time to failure is longer that it should be and the dev cycle time suffers.</p>

<p>I think this issue should add support for any levels of parallelism, not hard code it for any arbitrarily magic number of parallelism.</p>

<p><b>The issue should be renamed to "Support nested parallel stages"</b></p>

<p>I also agree that the "pipeline_mockup.png" in <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> looks like the natural way to render this and it should scale well. Rendering the X(1 for example) top level parallel stages different then the N-X inner levels, as I have seen proposed, seems like a hack tailored to some specific use case.</p>
</pre>
</details>

</ul>
</details>

<p>Our team is also hoping for a solution to this, the setup we have is similar to what <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shahafd" class="user-hover" rel="shahafd">shahafd</a> describes, starting with some sequential pre-work, then splitting into windows and linux build, then ideally further splitting several times down the pipeline (multiple tests that each is parameterized over OS and HW configurations and fully parallelizable).</p>

<p>Right now our pipeline takes three times as long as it should because we really like to have the user friendly visualization we get from Blue Ocean. This is of course not ideal because time to failure is longer that it should be and the dev cycle time suffers.</p>

<p>I think this issue should add support for any levels of parallelism, not hard code it for any arbitrarily magic number of parallelism.</p>

<p><b>The issue should be renamed to "Support nested parallel stages"</b></p>

<p>I also agree that the "pipeline_mockup.png" in <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> looks like the natural way to render this and it should scale well. Rendering the X(1 for example) top level parallel stages different then the N-X inner levels, as I have seen proposed, seems like a hack tailored to some specific use case.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rthakkar:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=367362&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367362">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>My set up is also very similar to the one that <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shahafd" class="user-hover" rel="shahafd">shahafd</a> has described.</p>

<p>A possible solution is to combine all nested stages into one node. When you click on this node, it expands and shows you everything else that was under it. This way the view is fairly clean and intuitive to use. See following images for examples:</p>

<p>In the image below lets imagine I clicked on build 1:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/47427/47427_Screen+Shot+2019-05-14+at+9.49.06+AM.png" height="304" width="169" style="border: 0px solid black" /></span></p>

<p>The UI will then transition to show the following nested pipeline:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/47428/47428_Screen+Shot+2019-05-20+at+8.55.46+PM.png" height="198" width="451" style="border: 0px solid black" /></span></p>

<p> </p>

<p>This basically steps away from only supporting nested parallels to just supporting nested pipelines.</p>

<p>Is there anyone who can be considered the owner of this issue? It would be great if we could come up with a plan and start executing. As  <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shahafd" class="user-hover" rel="shahafd">shahafd</a>  said, when you have 50+ builds in parallel, the UI is pretty much useless without nesting support.</p>
</pre>
</details>

</ul>
</details>

<p>My set up is also very similar to the one that <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shahafd" class="user-hover" rel="shahafd">shahafd</a> has described.</p>

<p>A possible solution is to combine all nested stages into one node. When you click on this node, it expands and shows you everything else that was under it. This way the view is fairly clean and intuitive to use. See following images for examples:</p>

<p>In the image below lets imagine I clicked on build 1:</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/47427/Screen%20Shot%202019-05-14%20at%209.49.06%20AM.png" height="304" width="169" style="border: 0px solid black" /></span></p>

<p>The UI will then transition to show the following nested pipeline:</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/47428/Screen%20Shot%202019-05-20%20at%208.55.46%20PM.png" height="198" width="451" style="border: 0px solid black" /></span></p>

<p> </p>

<p>This basically steps away from only supporting nested parallels to just supporting nested pipelines.</p>

<p>Is there anyone who can be considered the owner of this issue? It would be great if we could come up with a plan and start executing. As  <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shahafd" class="user-hover" rel="shahafd">shahafd</a>  said, when you have 50+ builds in parallel, the UI is pretty much useless without nesting support.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rthakkar:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=367363&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367363">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Could someone from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cloudbees" class="user-hover" rel="cloudbees">cloudbees</a> possibly take a look at this? I have seen a few other tickets related to parallel visualization in blue ocean owned by you guys.</p>
</pre>
</details>

</ul>
</details>

<p>Could someone from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cloudbees" class="user-hover" rel="cloudbees">cloudbees</a> possibly take a look at this? I have seen a few other tickets related to parallel visualization in blue ocean owned by you guys.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ianfixes:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=367505&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367505">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The essence of this issue is a mismatch between the ability to express a nested (and valid) workflow in the Jenkinsfile and the UI's ability to render it.  I don't know of a limit on nesting in the Jenkinsfile, so it follows that there shouldn't be such a limit in the UI either.</p>

<p> </p>

<p>To me, the significance of this issue is that BlueOcean is used for CI – its job is to help us execute, understand, and troubleshoot some complex procedures in dev/release.  Failing to represent the procedure itself undermines this entire concept.</p>
</pre>
</details>

</ul>
</details>

<p>The essence of this issue is a mismatch between the ability to express a nested (and valid) workflow in the Jenkinsfile and the UI's ability to render it.  I don't know of a limit on nesting in the Jenkinsfile, so it follows that there shouldn't be such a limit in the UI either.</p>

<p> </p>

<p>To me, the significance of this issue is that BlueOcean is used for CI – its job is to help us execute, understand, and troubleshoot some complex procedures in dev/release.  Failing to represent the procedure itself undermines this entire concept.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rthakkar:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=367506&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367506">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is it possible to bump the priority of this issue to a blocker or do we need someone special to do that?</p>
</pre>
</details>

</ul>
</details>

<p>Is it possible to bump the priority of this issue to a blocker or do we need someone special to do that?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rthakkar:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=367979&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367979">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I am bumping priority to blocker. Please let me know if this is something that I was not allowed to do.</p>

<p>Hopefully, this will bring a little more attention to this issue.</p>
</pre>
</details>

</ul>
</details>

<p>I am bumping priority to blocker. Please let me know if this is something that I was not allowed to do.</p>

<p>Hopefully, this will bring a little more attention to this issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>luckyhorang:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=371200&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-371200">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I am looking forward parallel in parallel feature, whole code of declarative syntax.</p>
</pre>
</details>

</ul>
</details>

<p>I am looking forward parallel in parallel feature, whole code of declarative syntax.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>trymb:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=373353&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-373353">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This is a very important issue for us to be resolved. Without this our pipeline takes so much longer to complete, since almost everything has to be done in series instead of parallel. </p>
</pre>
</details>

</ul>
</details>

<p>This is a very important issue for us to be resolved. Without this our pipeline takes so much longer to complete, since almost everything has to be done in series instead of parallel. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jeremfg.png" title="jeremfg's avatar" /> jeremfg:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=379619&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-379619">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm surprised this isn't supported by the UI.</p>

<p>We develop firmware and cross-compile to different architecture using agent of different OS. This equates to a lot of builds in parallels, with internal stages in parallel as well (such as static code analysis, running tests, etc.).</p>

<p>Parallelism is a must otherwise build times are way too long and a pain for developers who want rapid cycles of feedback and not wait for a build to complete.</p>
</pre>
</details>

</ul>
</details>

<p>I'm surprised this isn't supported by the UI.</p>

<p>We develop firmware and cross-compile to different architecture using agent of different OS. This equates to a lot of builds in parallels, with internal stages in parallel as well (such as static code analysis, running tests, etc.).</p>

<p>Parallelism is a must otherwise build times are way too long and a pain for developers who want rapid cycles of feedback and not wait for a build to complete.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>shahafd:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=379683&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-379683">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is there any reason why this issue doesn't get prioritized ?</p>

<p>It is already marked as a blocker for 6 months, and was critical for another 6 months earlier.</p>

<p>Can you please update on the status of this request ? Is it in development ? Do you plan to add this feature soon ?</p>

<p>It is very crucial for many people using BlueOcean plugin.</p>
</pre>
</details>

</ul>
</details>

<p>Is there any reason why this issue doesn't get prioritized ?</p>

<p>It is already marked as a blocker for 6 months, and was critical for another 6 months earlier.</p>

<p>Can you please update on the status of this request ? Is it in development ? Do you plan to add this feature soon ?</p>

<p>It is very crucial for many people using BlueOcean plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ianfixes:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=379692&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-379692">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I think a better question to ask would be "what technical hurdles block this request".  My guess is that when you run things in parallel, each parallel item counts as a "stage" to the UI, and so what we are really discussing here is a need to support sub-stages – something that seems a bit more complicated than a UI fix.</p>

<p>Is that accurate? </p>
</pre>
</details>

</ul>
</details>

<p>I think a better question to ask would be "what technical hurdles block this request".  My guess is that when you run things in parallel, each parallel item counts as a "stage" to the UI, and so what we are really discussing here is a need to support sub-stages – something that seems a bit more complicated than a UI fix.</p>

<p>Is that accurate? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rpuvvada:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=380423&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-380423">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>+1 I'm currently running 20 parallel stages in my pipeline for speed but the BlueOcean visualization of a tall column of nodes isn't the most user friendly. Nested parallel would be one solution to this. I'd also be happy with a purely UI construct to categorize my stages for better visualization. This has been mentioned in a series of tickets resolved as dupes that ultimately end up here. An example based on the description:</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
    agent none
    stages {
        stage(<span class="code-quote">'Parallel Stage'</span>) {
            parallel {
                stage(<span class="code-quote">'Stage 1'</span>) {
                    steps {
                        echo <span class="code-quote">"Stage 1"</span>
                    }
                }
                category(<span class="code-quote">'Stage 2'</span>) {
                    stage(<span class="code-quote">"Stage 2.1"</span>): {
                        steps {
                            echo <span class="code-quote">"Stage 2.1."</span>
                        }
                    },
                    stage(Stage 2.2") {
                        steps {
                            echo <span class="code-quote">"Stage 2.2."</span>
                        }
                    }
                }
            }
        }
    }
}
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>+1 I'm currently running 20 parallel stages in my pipeline for speed but the BlueOcean visualization of a tall column of nodes isn't the most user friendly. Nested parallel would be one solution to this. I'd also be happy with a purely UI construct to categorize my stages for better visualization. This has been mentioned in a series of tickets resolved as dupes that ultimately end up here. An example based on the description:</p>

<p> </p>

<pre>
pipeline {
    agent none
    stages {
stage(<span class="code-quote">'Parallel Stage'</span>) {
    parallel {
stage(<span class="code-quote">'Stage 1'</span>) {
    steps {
echo <span class="code-quote">"Stage 1"</span>
    }
}
category(<span class="code-quote">'Stage 2'</span>) {
    stage(<span class="code-quote">"Stage 2.1"</span>): {
steps {
    echo <span class="code-quote">"Stage 2.1."</span>
}
    },
    stage(Stage 2.2") {
steps {
    echo <span class="code-quote">"Stage 2.2."</span>
}
    }
}
    }
}
    }
}
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rockmax:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=381959&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381959">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Blueocean team - Any updates on this? Are you guys working on this issue since its a blocker for us?</p>

<p> </p>

<ul class="alternate" type="square">
	<li>Adobe Developer</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>Blueocean team - Any updates on this? Are you guys working on this issue since its a blocker for us?</p>

<p> </p>

<ul class="alternate" type="square">
	<li>Adobe Developer</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>shanidar1:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=381966&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381966">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>What I did as a workaround is building a parallel stage dynamically - </p>

<p>We deploy 5 services for two clouds, we want to deploy all in parallel.  Let's say each service takes 5 min to deploy, we had to wait 10 minutes because we couldn't run two cloud stages in parallel that runs multiple services stages also in parallel:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/49810/49810_Selection_093.png" style="border: 0px solid black" /></span></p>

<p>So I built one parallel stage with all 10 services, and named each step with a relevant prefix:</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">'Deploy'</span>) {
                    steps {
                        script {
                            def parallelStagesMap = [:]
                                AWS_CLUSTER_NAME.collect {
                                   parallelStagesMap.put(<span class="code-quote">"AWS ${it}"</span>, buildUtils.Deploy(<span class="code-quote">'AWS'</span>, it))
                                }
                                GCP_CLUSTER_NAME.collect {
                                   parallelStagesMap.put(<span class="code-quote">"GCP ${it}"</span>, buildUtils.Deploy(<span class="code-quote">'GCP'</span>, it))
                                }

                             parallel parallelStagesMap
                        }
                    }
                }
</pre>
</div></div>
<p> </p>

<p> </p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/49809/49809_Selection_090.png" style="border: 0px solid black" /></span></p>

<p> </p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>What I did as a workaround is building a parallel stage dynamically - </p>

<p>We deploy 5 services for two clouds, we want to deploy all in parallel.  Let's say each service takes 5 min to deploy, we had to wait 10 minutes because we couldn't run two cloud stages in parallel that runs multiple services stages also in parallel:</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/49810/Selection_093.png" style="border: 0px solid black" /></span></p>

<p>So I built one parallel stage with all 10 services, and named each step with a relevant prefix:</p>

<p> </p>

<pre>
stage(<span class="code-quote">'Deploy'</span>) {
    steps {
script {
    def parallelStagesMap = [:]
AWS_CLUSTER_NAME.collect {
   parallelStagesMap.put(<span class="code-quote">"AWS ${it}"</span>, buildUtils.Deploy(<span class="code-quote">'AWS'</span>, it))
}
GCP_CLUSTER_NAME.collect {
   parallelStagesMap.put(<span class="code-quote">"GCP ${it}"</span>, buildUtils.Deploy(<span class="code-quote">'GCP'</span>, it))
}

     parallel parallelStagesMap
}
    }
}
</pre>
<p> </p>

<p> </p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/49809/Selection_090.png" style="border: 0px solid black" /></span></p>

<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ianfixes:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=381996&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381996">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>That's a good workaround.  I can't speak for everyone here, but my personal interest in this issue is more to do with the first attached screenshot in the original issue &#8211; the ability to have multiple sequential stages shown in each parallel branch.  In other words, to not have it all squish horizontally into a single stage that represents the entire branch.</p>
</pre>
</details>

</ul>
</details>

<p>That's a good workaround.  I can't speak for everyone here, but my personal interest in this issue is more to do with the first attached screenshot in the original issue &#8211; the ability to have multiple sequential stages shown in each parallel branch.  In other words, to not have it all squish horizontally into a single stage that represents the entire branch.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>shanidar1:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=381998&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381998">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ianfixes" class="user-hover" rel="ianfixes">ianfixes</a> I agree, I'm also waiting for this solution, that's what I did meanwhile until this issue will be resolved..</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ianfixes" class="user-hover" rel="ianfixes">ianfixes</a> I agree, I'm also waiting for this solution, that's what I did meanwhile until this issue will be resolved..</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rvnairme:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=386064&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-386064">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is there a plan to update the visualization as shown in parallel-stages-new.png ? Or is it available in already in any of the new releases ?</p>
</pre>
</details>

</ul>
</details>

<p>Is there a plan to update the visualization as shown in parallel-stages-new.png ? Or is it available in already in any of the new releases ?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amidar.jpg" title="amidar's avatar" /> amidar:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=392615&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-392615">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi,</p>

<p> </p>

<p>any update regarding this issue?</p>
</pre>
</details>

</ul>
</details>

<p>Hi,</p>

<p> </p>

<p>any update regarding this issue?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>henryborchers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=398778&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-398778">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Having multiple levels of parallel stages would optimize the use of my executors and significantly reduce my pipeline times. </p>

<p>This would be very helpful to me for stages that create multiple platform specific artifacts. When I do this now, I have to have one matrix stage that builds Windows and Linux packages (using creative use of Dockerfile agents, I'm able to get these to run in parallel), and one stage dedicated to building Mac packages. Because matrix stages seem to use up a parallel stage under the hood, I can only run these stages one after another, even though the physical servers could easily run these steps independently .  </p>
</pre>
</details>

</ul>
</details>

<p>Having multiple levels of parallel stages would optimize the use of my executors and significantly reduce my pipeline times. </p>

<p>This would be very helpful to me for stages that create multiple platform specific artifacts. When I do this now, I have to have one matrix stage that builds Windows and Linux packages (using creative use of Dockerfile agents, I'm able to get these to run in parallel), and one stage dedicated to building Mac packages. Because matrix stages seem to use up a parallel stage under the hood, I can only run these stages one after another, even though the physical servers could easily run these steps independently .  </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>alonmiz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=399142&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-399142">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>that's a really important feature for our team as well. we have an optimized ci with multiple parallelized steps, which look really awkward without that capability</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=henryborchers" class="user-hover" rel="henryborchers">henryborchers</a> in your case there's no limitation in using parallelism inside another. the downside is that it's really hard to view the process in the UI and in case of a failure step, the UI just don't understand how to show it.</p>
</pre>
</details>

</ul>
</details>

<p>that's a really important feature for our team as well. we have an optimized ci with multiple parallelized steps, which look really awkward without that capability</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=henryborchers" class="user-hover" rel="henryborchers">henryborchers</a> in your case there's no limitation in using parallelism inside another. the downside is that it's really hard to view the process in the UI and in case of a failure step, the UI just don't understand how to show it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>djviking:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=402277&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-402277">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Will this fix the problem of visually showing two parallel Pipelines we have now run into?</p>

<p>We have a Jenkinsfile which runs two Pipelines in serial. For efficiency these can be run in parallel, but that will not look good din Blue Ocean.
This is how the two Pipelines are visualized in Blue Ocean (The black circle is the first pipeline, the blue is the second).
 <span class="image-wrap" style=""><a id="53590_thumb" href="https://issues.jenkins.io/secure/attachment/53590/53590_TwoPipelinesInSerial.png" title="TwoPipelinesInSerial.png" file-preview-type="image" file-preview-id="53590" file-preview-title="TwoPipelinesInSerial.png"><img src="https://issues.jenkins.io/secure/thumbnail/53590/_thumb_53590.png" style="border: 0px solid black" role="presentation"/></a></span> 
This is how the two Pipelines are visualized when running them in parallel:
 <span class="image-wrap" style=""><a id="53591_thumb" href="https://issues.jenkins.io/secure/attachment/53591/53591_TwoPipelinesInParallel.png" title="TwoPipelinesInParallel.png" file-preview-type="image" file-preview-id="53591" file-preview-title="TwoPipelinesInParallel.png"><img src="https://issues.jenkins.io/secure/thumbnail/53591/_thumb_53591.png" style="border: 0px solid black" role="presentation"/></a></span> </p>

<p>The Scripted Pipeline in Seriel:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-groovy"><span class="code-keyword">def</span> maven = <span class="code-keyword">new</span> com.company.MavenBuild()
maven.executeBuild()
<span class="code-keyword">def</span> gradle = <span class="code-keyword">new</span> com.company.GradleBuild()
gradle.executeBuild()
</pre>
</div></div>

<p>The Scripted Pipeline in Parallel:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-groovy">parallel {
    <span class="code-quote">'Maven'</span>: {
        <span class="code-keyword">def</span> maven = <span class="code-keyword">new</span> com.company.MavenBuild()
        maven.executeBuild()
    },
    <span class="code-quote">'Gradle'</span>: {
        <span class="code-keyword">def</span> gradle = <span class="code-keyword">new</span> com.company.GradleBuild()
        gradle.executeBuild()
    }
}
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>Will this fix the problem of visually showing two parallel Pipelines we have now run into?</p>

<p>We have a Jenkinsfile which runs two Pipelines in serial. For efficiency these can be run in parallel, but that will not look good din Blue Ocean.<br/>
This is how the two Pipelines are visualized in Blue Ocean (The black circle is the first pipeline, the blue is the second).<br/>
 <span class="image-wrap" style=""><a id="53590_thumb" href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/53590/TwoPipelinesInSerial.png" title="TwoPipelinesInSerial.png" file-preview-type="image" file-preview-id="53590" file-preview-title="TwoPipelinesInSerial.png"><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/53590/TwoPipelinesInSerial.png" style="border: 0px solid black" role="presentation"/></a></span> <br/>
This is how the two Pipelines are visualized when running them in parallel:<br/>
 <span class="image-wrap" style=""><a id="53591_thumb" href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/53591/TwoPipelinesInParallel.png" title="TwoPipelinesInParallel.png" file-preview-type="image" file-preview-id="53591" file-preview-title="TwoPipelinesInParallel.png"><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/53591/TwoPipelinesInParallel.png" style="border: 0px solid black" role="presentation"/></a></span> </p>

<p>The Scripted Pipeline in Seriel:</p>

<pre>
<span class="code-keyword">def</span> maven = <span class="code-keyword">new</span> com.company.MavenBuild()
maven.executeBuild()
<span class="code-keyword">def</span> gradle = <span class="code-keyword">new</span> com.company.GradleBuild()
gradle.executeBuild()
</pre>

<p>The Scripted Pipeline in Parallel:</p>

<pre>
parallel {
    <span class="code-quote">'Maven'</span>: {
<span class="code-keyword">def</span> maven = <span class="code-keyword">new</span> com.company.MavenBuild()
maven.executeBuild()
    },
    <span class="code-quote">'Gradle'</span>: {
<span class="code-keyword">def</span> gradle = <span class="code-keyword">new</span> com.company.GradleBuild()
gradle.executeBuild()
    }
}
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/llibicpep.jpg" title="llibicpep's avatar" /> llibicpep:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=403528&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-403528">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Here's a use case that might help with design.</p>

<p>Consider a typical pipeline that build an artifact<span class="error">&#91;s&#93;</span> and deploys it to sequential environments (let's say Dev and QA for the sake of example).</p>

<p>Consider that you might want to deploy to multiple locations, for instance - to aws, azure and gcp. These are parallel independent worlds, so environments progression would be happening in parallel there.</p>

<p>Consider that within each location, you also want to deploy into multiple regions. Regions are also isolated independent worlds, so they would have to be deployed in parallel within it's location (aka aws, gcp, azure etc).</p>

<p>That creates at least two levels of nesting, but generally you can extrapolate this idea, for instance to deploy to each AZ separately, or maybe - deploy to multiple groups of locations or something like that, I do not see why there should be a limit of just two.</p>

<p>Anyway, here is a scripted pipeline representation of that use case (just two levels for the sake of simplicity):</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"> stage(<span class="code-quote">'Realms'</span>) {
    parallel <span class="code-quote">'aws'</span>: {
        stage(<span class="code-quote">'Regions (aws)'</span>) {
            parallel <span class="code-quote">'us-east-1 (aws)'</span>: {
                stage(<span class="code-quote">'Dev (aws us-east-1)'</span>) {
                    echo <span class="code-quote">'aws us-east-1 dev'</span>
                }
                stage(<span class="code-quote">'QA (aws us-east-1)'</span>) {
                    input message: <span class="code-quote">'Approve aws us-east-1 qa?'</span>, ok: <span class="code-quote">'Approve aws us-east-1 qa!'</span>
                    echo <span class="code-quote">'aws us-east-1 qa'</span>
                }
            }, <span class="code-quote">'us-west-2 (aws)'</span>: {
                stage(<span class="code-quote">'Dev (aws us-west-2)'</span>) {
                    echo <span class="code-quote">'aws us-west-2 dev'</span>
                }
                stage(<span class="code-quote">'QA (aws us-west-2)'</span>) {
                    input message: <span class="code-quote">'Approve aws us-west-2 qa?'</span>, ok: <span class="code-quote">'Approve aws us-west-2 qa!'</span>
                    echo <span class="code-quote">'aws us-west-2 qa'</span>
                }
            }, failFast: <span class="code-keyword">false</span>
        }
    }, <span class="code-quote">'gcp'</span>: {
        stage(<span class="code-quote">'Regions (gcp)'</span>) {
            parallel <span class="code-quote">'us-east-1 (gcp)'</span>: {
                stage(<span class="code-quote">'Dev (gcp us-east4)'</span>) {
                    echo <span class="code-quote">'gcp us-east4 dev'</span>
                }
                stage(<span class="code-quote">'QA (gcp us-east4)'</span>) {
                    input message: <span class="code-quote">'Approve gcp us-east4 qa?'</span>, ok: <span class="code-quote">'Approve gcp us-east4 qa!'</span>
                    echo <span class="code-quote">'gcp us-east4 qa'</span>
                }
            }, <span class="code-quote">'us-west-2 (gcp)'</span>: {
                stage(<span class="code-quote">'Dev (gcp us-west4)'</span>) {
                    echo <span class="code-quote">'gcp us-west4 dev'</span>
                }
                stage(<span class="code-quote">'QA (gcp us-west4)'</span>) {
                    input message: <span class="code-quote">'Approve gcp us-west4 qa?'</span>, ok: <span class="code-quote">'Approve gcp us-west4 qa!'</span>
                    echo <span class="code-quote">'gcp us-west4 qa'</span>
                }
            }, failFast: <span class="code-keyword">false</span>
        }
    }, failFast: <span class="code-keyword">false</span>
}
</pre>
</div></div>

<p>Also consider a deployment stage itself (currently represented with just one simple echo step) - it will most likely have a lot of low level steps such as sh and so on. From the  pipeline user perspective - it doesn't makes any sense to see all that implementation details. The user will want to see aggregation of steps at some level of abstractions, for instance if deployment is implemented with Terraform - the user will want to see things like plan, display diff and apply. The user might want to dig into any of these abstractions to the level of actual steps, of course, so that should be an expandable option.
I think implementing more than two levels of nesting using a use case from above will also solve that second use case.</p>
</pre>
</details>

</ul>
</details>

<p>Here's a use case that might help with design.</p>

<p>Consider a typical pipeline that build an artifact<span class="error">&#91;s&#93;</span> and deploys it to sequential environments (let's say Dev and QA for the sake of example).</p>

<p>Consider that you might want to deploy to multiple locations, for instance - to aws, azure and gcp. These are parallel independent worlds, so environments progression would be happening in parallel there.</p>

<p>Consider that within each location, you also want to deploy into multiple regions. Regions are also isolated independent worlds, so they would have to be deployed in parallel within it's location (aka aws, gcp, azure etc).</p>

<p>That creates at least two levels of nesting, but generally you can extrapolate this idea, for instance to deploy to each AZ separately, or maybe - deploy to multiple groups of locations or something like that, I do not see why there should be a limit of just two.</p>

<p>Anyway, here is a scripted pipeline representation of that use case (just two levels for the sake of simplicity):</p>


<pre>
 stage(<span class="code-quote">'Realms'</span>) {
    parallel <span class="code-quote">'aws'</span>: {
stage(<span class="code-quote">'Regions (aws)'</span>) {
    parallel <span class="code-quote">'us-east-1 (aws)'</span>: {
stage(<span class="code-quote">'Dev (aws us-east-1)'</span>) {
    echo <span class="code-quote">'aws us-east-1 dev'</span>
}
stage(<span class="code-quote">'QA (aws us-east-1)'</span>) {
    input message: <span class="code-quote">'Approve aws us-east-1 qa?'</span>, ok: <span class="code-quote">'Approve aws us-east-1 qa!'</span>
    echo <span class="code-quote">'aws us-east-1 qa'</span>
}
    }, <span class="code-quote">'us-west-2 (aws)'</span>: {
stage(<span class="code-quote">'Dev (aws us-west-2)'</span>) {
    echo <span class="code-quote">'aws us-west-2 dev'</span>
}
stage(<span class="code-quote">'QA (aws us-west-2)'</span>) {
    input message: <span class="code-quote">'Approve aws us-west-2 qa?'</span>, ok: <span class="code-quote">'Approve aws us-west-2 qa!'</span>
    echo <span class="code-quote">'aws us-west-2 qa'</span>
}
    }, failFast: <span class="code-keyword">false</span>
}
    }, <span class="code-quote">'gcp'</span>: {
stage(<span class="code-quote">'Regions (gcp)'</span>) {
    parallel <span class="code-quote">'us-east-1 (gcp)'</span>: {
stage(<span class="code-quote">'Dev (gcp us-east4)'</span>) {
    echo <span class="code-quote">'gcp us-east4 dev'</span>
}
stage(<span class="code-quote">'QA (gcp us-east4)'</span>) {
    input message: <span class="code-quote">'Approve gcp us-east4 qa?'</span>, ok: <span class="code-quote">'Approve gcp us-east4 qa!'</span>
    echo <span class="code-quote">'gcp us-east4 qa'</span>
}
    }, <span class="code-quote">'us-west-2 (gcp)'</span>: {
stage(<span class="code-quote">'Dev (gcp us-west4)'</span>) {
    echo <span class="code-quote">'gcp us-west4 dev'</span>
}
stage(<span class="code-quote">'QA (gcp us-west4)'</span>) {
    input message: <span class="code-quote">'Approve gcp us-west4 qa?'</span>, ok: <span class="code-quote">'Approve gcp us-west4 qa!'</span>
    echo <span class="code-quote">'gcp us-west4 qa'</span>
}
    }, failFast: <span class="code-keyword">false</span>
}
    }, failFast: <span class="code-keyword">false</span>
}
</pre>

<p>Also consider a deployment stage itself (currently represented with just one simple echo step) - it will most likely have a lot of low level steps such as sh and so on. From the  pipeline user perspective - it doesn't makes any sense to see all that implementation details. The user will want to see aggregation of steps at some level of abstractions, for instance if deployment is implemented with Terraform - the user will want to see things like plan, display diff and apply. The user might want to dig into any of these abstractions to the level of actual steps, of course, so that should be an expandable option.<br/>
I think implementing more than two levels of nesting using a use case from above will also solve that second use case.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gmtm:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=409817&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-409817">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Does anyone know if this is purely a visualization issue? or it is an issue with the underlying data structure? If someone was wanting to get started looking into this issue, where should they start?</p>
</pre>
</details>

</ul>
</details>

<p>Does anyone know if this is purely a visualization issue? or it is an issue with the underlying data structure? If someone was wanting to get started looking into this issue, where should they start?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>henryborchers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=410861&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-410861">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=gmtm" class="user-hover" rel="gmtm">gmtm</a>, I get the feeling the project does not have anybody working on Blue Ocean development anymore. My guess is that whoever was in charge of it has left the company or moved on to other projects because I haven't seen any real improvements done to it in years. It's a shame because Blue Ocean was a GAME CHANGER!!! I personally love how Blue Ocean displays my pipeline and helps me quickly figure out an issue when something goes wrong.  However, it seems to have issues that don't aren't being addressed and missing features (like this one) that could make it much. (In addition, Blue Ocean takes a very long time to load!)</p>

<p>I would LOVE first-class support for multiple levels of parallelism (such as nested parallel commands or nested matrixes in the declarative pipelines). Having that would let me load-balance and greatly reduce the duration of my pipelines. However, it just doesn't seem like a priority for anybody who has the skills to do anything about it.</p>

<p>I don't blame anybody working on Jenkins. In fact, I love everything they do regardless if they address stuff that make my life better or someone else's. No, I just wish my own skills included what was needed to help out in the same way the Jenkins has helped me. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=gmtm" class="user-hover" rel="gmtm">gmtm</a>, I get the feeling the project does not have anybody working on Blue Ocean development anymore. My guess is that whoever was in charge of it has left the company or moved on to other projects because I haven't seen any real improvements done to it in years. It's a shame because Blue Ocean was a GAME CHANGER!!! I personally love how Blue Ocean displays my pipeline and helps me quickly figure out an issue when something goes wrong.  However, it seems to have issues that don't aren't being addressed and missing features (like this one) that could make it much. (In addition, Blue Ocean takes a very long time to load!)</p>

<p>I would LOVE first-class support for multiple levels of parallelism (such as nested parallel commands or nested matrixes in the declarative pipelines). Having that would let me load-balance and greatly reduce the duration of my pipelines. However, it just doesn't seem like a priority for anybody who has the skills to do anything about it.</p>

<p>I don't blame anybody working on Jenkins. In fact, I love everything they do regardless if they address stuff that make my life better or someone else's. No, I just wish my own skills included what was needed to help out in the same way the Jenkins has helped me. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/llibicpep.jpg" title="llibicpep's avatar" /> llibicpep:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=411124&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-411124">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Well said <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=henryborchers" class="user-hover" rel="henryborchers">henryborchers</a>. Additionally I would point out that classic UI does not have visualization support for parallel stages at all, in any shape or form, not even single leveled. It is all unreadable linear mess in the classic UI.</p>
</pre>
</details>

</ul>
</details>

<p>Well said <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=henryborchers" class="user-hover" rel="henryborchers">henryborchers</a>. Additionally I would point out that classic UI does not have visualization support for parallel stages at all, in any shape or form, not even single leveled. It is all unreadable linear mess in the classic UI.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stuartrowe.png" title="stuartrowe's avatar" /> stuartrowe:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=411126&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-411126">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There hasn't been any active development on Blue Ocean for a while. See <a href="https://groups.google.com/g/jenkinsci-users/c/xngZrSsXIjc/m/-_aIrWyICgAJ." class="external-link" target="_blank" rel="nofollow noopener">https://groups.google.com/g/jenkinsci-users/c/xngZrSsXIjc/m/-_aIrWyICgAJ.</a></p>

<p>The <a href="https://github.com/jenkinsci/pipeline-graph-view-plugin" class="external-link" target="_blank" rel="nofollow noopener">pipeline-graph-view-plugin</a> is bringing the pipeline visualization from Blue Ocean into the classic UI. It might be worth creating a new issue for that plugin?</p>
</pre>
</details>

</ul>
</details>

<p>There hasn't been any active development on Blue Ocean for a while. See <a href="https://groups.google.com/g/jenkinsci-users/c/xngZrSsXIjc/m/-_aIrWyICgAJ." class="external-link" target="_blank" rel="nofollow noopener">https://groups.google.com/g/jenkinsci-users/c/xngZrSsXIjc/m/-_aIrWyICgAJ.</a></p>

<p>The <a href="https://github.com/jenkinsci/pipeline-graph-view-plugin" class="external-link" target="_blank" rel="nofollow noopener">pipeline-graph-view-plugin</a> is bringing the pipeline visualization from Blue Ocean into the classic UI. It might be worth creating a new issue for that plugin?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gl1koz3:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=416687&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-416687">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>+100500 Jenkins shared libraries and Jenkinsfiles are awesome, but the UI around them is a terrible mess, especially when it comes to stage display and generally trying to understand what logs are where etc. I think it's easier to rewrite the UI for pipeline builds from scratch to somehow allow browsing through everything in a visually pleasing execution step graph, not a flat representation like a list or a table.</p>
</pre>
</details>

</ul>
</details>

<p>+100500 Jenkins shared libraries and Jenkinsfiles are awesome, but the UI around them is a terrible mess, especially when it comes to stage display and generally trying to understand what logs are where etc. I think it's easier to rewrite the UI for pipeline builds from scratch to somehow allow browsing through everything in a visually pleasing execution step graph, not a flat representation like a list or a table.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dkiaunis:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54010?focusedId=423112&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-423112">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is this still being worked on?</p>
</pre>
</details>

</ul>
</details>

<p>Is this still being worked on?</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-50532%22">JENKINS-50532</a>]</i>

**Comment by timja** [comment]
This feature has been added to Pipeline Graph View the successor of BlueOcean.

See the release notes in https://github.com/jenkinsci/pipeline-graph-view-plugin/releases/tag/896.v29722831939d
